infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java

```diff
@@ -0,0 +1,117 @@
+                                Layout.LayoutSegment segment) throws ExecutionException, InterruptedException {
+
+        int batchSize = runtime.getParameters().getBulkReadSize();
+
+        long trimMark = runtime.getAddressSpaceView().getTrimMark().getSequence();
+        // Send the trimMark to the new/healing nodes.
+        // If this times out or fails, the Action performing the stateTransfer fails and retries.
+        for (String endpoint : endpoints) {
+            // TrimMark is the first address present on the log unit server.
+            // Perform the prefix trim on the preceding address = (trimMark - 1).
+            // Since the LU will reject trim decisions made from older epochs, we
+            // need to adjust the new trim mark to have the new layout's epoch.
+            Token prefixToken = new Token(layout.getEpoch(), trimMark - 1);
+            CFUtils.getUninterruptibly(runtime.getLayoutView().getRuntimeLayout(layout)
+                    .getLogUnitClient(endpoint)
+                    .prefixTrim(prefixToken));
+        }
+
+        if (trimMark > segment.getEnd()) {
+            log.info("stateTransfer: Nothing to transfer, trimMark {} greater than end of segment {}",
+                    trimMark, segment.getEnd());
+            return;
+        }
+
+        // State transfer should start from segment start address or trim mark whichever is lower.
+        long segmentStart = Math.max(trimMark, segment.getStart());
+
+        for (long chunkStart = segmentStart; chunkStart < segment.getEnd()
+                ; chunkStart = chunkStart + batchSize) {
+            long chunkEnd = Math.min((chunkStart + batchSize - 1), segment.getEnd() - 1);
+
+            long ts1 = System.currentTimeMillis();
+
+            Map<Long, ILogData> dataMap = runtime.getAddressSpaceView()
+                    .cacheFetch(ContiguousSet.create(
+                            Range.closed(chunkStart, chunkEnd),
+                            DiscreteDomain.longs()));
+
+            long ts2 = System.currentTimeMillis();
+
+            log.info("stateTransfer: read {}-{} in {} ms", chunkStart, chunkEnd, (ts2 - ts1));
+
+            List<LogData> entries = new ArrayList<>();
+            for (long x = chunkStart; x <= chunkEnd; x++) {
+                if (dataMap.get(x) == null) {
+                    log.error("Missing address {} in range {}-{}", x, chunkStart, chunkEnd);
+                    throw new IllegalStateException("Missing address");
+                }
+                entries.add((LogData) dataMap.get(x));
+            }
+
+            for (String endpoint : endpoints) {
+                // Write segment chunk to the new logunit
+                ts1 = System.currentTimeMillis();
+                boolean transferSuccess = runtime.getLayoutView().getRuntimeLayout(layout)
+                        .getLogUnitClient(endpoint)
+                        .writeRange(entries).get();
+                ts2 = System.currentTimeMillis();
```

Comment:
can we return completablefuture instead of blocking current thread and waiting for result?

Reply:
So, if you don't start using it properly, you will always use `get()`, why not to start using it properly right now?
You can easily measure performance for this, just execute `whenComplete()` and measure the time taken for this operation.